File: 152-maximum-product-subarray/152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        # kadane max(max_pre * curr , cur)
        
        ans = max(nums) ; mincurr = nums[0] ; maxcurr = nums[0]
        for i in range(1,len(nums)):
            curr = nums[i]
            temp_mincurr = min(curr,mincurr * curr , maxcurr * curr)
            maxcurr = max(curr,maxcurr * curr , mincurr * curr)
            mincurr = temp_mincurr
            ans = max(ans,maxcurr)
        return ans

File: 152-maximum-product-subarray/152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int mincurr = nums[0] ; 
        int maxcurr = nums[0] ; 
        int ans = maxcurr ; 
        
        for (int i = 1 ; i < nums.length ;i++){
            int curr = nums[i];
            int temp_mincurr = Math.min(curr,Math.min(curr * mincurr,curr * maxcurr));
            maxcurr = Math.max(curr,Math.max(curr * mincurr,curr * maxcurr));
            mincurr = temp_mincurr;
            ans = Math.max(maxcurr,ans);
        }
        return ans;
        
    }
}